pageserver/src/tenant/timeline/layer_manager.rs

```diff
@@ -241,7 +269,9 @@ impl LayerManager {
         //      map index without actually rebuilding the index.
         updates.remove_historic(desc);
         mapping.remove(layer);
-        layer.delete_on_drop();
+        if physical {
+            layer.delete_on_drop();
+        }
```

Comment:
Is this needed because the filename does not contain generation...?

But if we overwrite the existing layer file, don't we then risk all kinds of problems, including reading the wrong image?

Reply:
Perhaps including the generation number to the file name would be the simplest way around this, we would risk no problems.

Reply:
This part was indeed half-done, removed in 6a995dd43 pending the next PR.

I'm going to look at that next.  Putting the generation number in local files does make it simpler to reason about, but it's a pretty invasive thing (would have to do large numbers of renames on pageserver upgrade), although I think we can make it work without (I'll ping you)